server/server.go

```diff
@@ -0,0 +1,377 @@
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	node, err := s.peer.Get(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	respNode, err := nodeToPbNode(node)
+	if err != nil {
+		return nil, err
+	}
+
+	return &pb.GetNodeResponse{Node: respNode}, nil
+}
+
+func (s *ipfsLiteServer) GetNodes(req *pb.GetNodesRequest, srv pb.IpfsLite_GetNodesServer) error {
+	cids := make([]cid.Cid, len(req.GetCids()))
+	for i, cidString := range req.GetCids() {
+		cid, err := cid.Decode(cidString)
+		if err != nil {
+			return err
+		}
+		cids[i] = cid
+	}
+	ch := s.peer.GetMany(srv.Context(), cids)
+	for {
+		result, ok := <-ch
+		if ok == false {
+			break
+		} else {
+			resp := pb.GetNodesResponse{}
+			if result.Err != nil {
+				resp.Option = &pb.GetNodesResponse_Error{Error: result.Err.Error()}
+			} else {
+				node, err := nodeToPbNode(result.Node)
+				if err != nil {
+					resp.Option = &pb.GetNodesResponse_Error{Error: err.Error()}
+				} else {
+					resp.Option = &pb.GetNodesResponse_Node{Node: node}
+				}
+			}
+			srv.Send(&resp)
+		}
+	}
+	return nil
+}
+
+func (s *ipfsLiteServer) RemoveNode(ctx context.Context, req *pb.RemoveNodeRequest) (*pb.RemoveNodeResponse, error) {
+	cid, err := cid.Decode(req.GetCid())
+	if err != nil {
+		return nil, err
+	}
+	err = s.peer.Remove(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.RemoveNodeResponse{}, nil
+}
+
+func (s *ipfsLiteServer) RemoveNodes(ctx context.Context, req *pb.RemoveNodesRequest) (*pb.RemoveNodesResponse, error) {
+	cids := make([]cid.Cid, len(req.GetCids()))
+	for i, reqCid := range req.GetCids() {
+		cid, err := cid.Decode(reqCid)
+		if err != nil {
+			return nil, err
+		}
+		cids[i] = cid
+	}
+	err := s.peer.RemoveMany(ctx, cids)
+	if err != nil {
+		return nil, err
+	}
+	return &pb.RemoveNodesResponse{}, nil
+}
+
+func (s *ipfsLiteServer) ResolveLink(ctx context.Context, req *pb.ResolveLinkRequest) (*pb.ResolveLinkResponse, error) {
+	cid, err := cid.Decode(req.GetNodeCid())
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := s.peer.Get(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	link, remainingPath, err := node.ResolveLink(req.GetPath())
+	if err != nil {
+		return nil, err
+	}
+
+	respLink := pb.Link{
+		Name: link.Name,
+		Size: int64(link.Size),
+		Cid:  link.Cid.String(),
+	}
+
+	resp := pb.ResolveLinkResponse{
+		Link:          &respLink,
+		RemainingPath: remainingPath,
+	}
+
+	return &resp, nil
+}
+
+func (s *ipfsLiteServer) Tree(ctx context.Context, req *pb.TreeRequest) (*pb.TreeResponse, error) {
+	cid, err := cid.Decode(req.GetNodeCid())
+	if err != nil {
+		return nil, err
+	}
+
+	node, err := s.peer.Get(ctx, cid)
+	if err != nil {
+		return nil, err
+	}
+
+	paths := node.Tree(req.GetPath(), int(req.GetDepth()))
+
+	return &pb.TreeResponse{Paths: paths}, nil
+}
+
+func pbBlockToBlock(pbBlock *pb.Block) (blocks.Block, error) {
```

Comment:
@sanderpick here's a method that's used in `AddNode` and `AddNodes` that checks if the request includes a cid and uses it if so. At this time, I have a test that doesn't send a cid. This method runs with no problem and returns a Block created only from the raw data (which is some cbor node data). See my next comment in `AddNode` about an error.

Reply:
For context, here is the test code that creates the Node and calls the AddNode method:
```
func createNode(data map[string]interface{}) (*cbor.Node, error) {
	codec := uint64(multihash.SHA2_256)
	return cbor.WrapObject(data, codec, multihash.DefaultLengths[codec])
}

func TestAddNode(t *testing.T) {
	...
	node0data := map[string]interface{}{
		"name": "node0",
	}
	node0, err := createNode(node0data)
	if err != nil {
		t.Fatalf("failed to create node: %v", err)
	}

	block0 := pb.Block{
		// Cid:     node0.Cid().String(),
		RawData: node0.RawData(),
	}

	_, err = client.AddNode(ctx, &pb.AddNodeRequest{Block: &block0})
	if err != nil {
		t.Fatalf("failed to add node: %v", err)
	}
	...
}
```

Reply:
Hmm, I could be mistaken about the block creation if it's not working. I guess RawData is serialized cbor, meaning there's nothing about it the self-declares it _is_ cbor, like a multicodec.

Reply:
What is the IPFS Lite method you're exposing here?

Reply:
This would be the `AddNode` and `AddNodes` methods

Reply:
So @sanderpick, I guess the bytes themselves aren't self describing -- it's the cid that describes how the bytes are encoded. I guess if you don't provide a cid, the default assumption is the bytes are Protobuf?

Reply:
Right, seems that way.

Reply:
Can you point me to the AddNode method on the ipfs lite repo?

Reply:
The typical way to add nodes if with a dag service, which the ipfs lite instance itself becomes. Then `Add` and `AddMany` are at your disposal. 

https://github.com/ipfs/go-ipld-format/blob/master/merkledag.go#L30

Reply:
Yea those are the exact methods I'm exposing.

Reply:
Ok, so if you `AddNode` or `AddNodes` without passing a Cid, the node data must be ProtoNode. If you pass a Cid, then you're free to use cbor or Protobuf or whatever supported format.